sdk/java-sdk-protobuf/src/main/scala/kalix/javasdk/impl/action/ActionEffectImpl.scala

```diff
@@ -101,13 +100,26 @@ object ActionEffectImpl {
     def asyncReply[S](futureMessage: CompletionStage[S]): Action.Effect[S] =
       asyncReply(futureMessage, Metadata.EMPTY)
     def asyncReply[S](futureMessage: CompletionStage[S], metadata: Metadata): Action.Effect[S] =
-      AsyncEffect(futureMessage.asScala.map(s => Builder.reply[S](s, metadata))(ExecutionContext.parasitic), Nil)
+      AsyncEffect(futureMessage.asScala.map(s => reply[S](s, metadata.addTracing()))(ExecutionContext.parasitic), Nil)
     def asyncEffect[S](futureEffect: CompletionStage[Action.Effect[S]]): Action.Effect[S] =
       AsyncEffect(futureEffect.asScala, Nil)
     def ignore[S](): Action.Effect[S] =
       IgnoreEffect()
+
+    import scala.jdk.OptionConverters._
+
+    implicit class TracingWrapper(metadata: Metadata) {
+      def addTracing(): Metadata = {
+        actionContextMetadata.traceContext().traceParent().toScala match {
+          case Some(traceparent) if !metadata.has(Telemetry.TRACE_PARENT_KEY) =>
+            metadata.add(Telemetry.TRACE_PARENT_KEY, traceparent)
+          case _ => metadata
+        }
+      }
+    }
+
   }
 
-  def builder(): Action.Effect.Builder = Builder
+  def builder(context: Metadata): Action.Effect.Builder = new Builder(context)
```

Comment:
Is this method in use? It seems not, because signature changed and there is no call to this method in this diff. So probably it's never called at all.

On the other hand, we can use it instead of `new ActionEffectImpl.Builder(Metadata.EMPTY)`.

Reply:
Yes, this is used [here](https://github.com/lightbend/kalix-jvm-sdk/pull/2172/files#diff-42fba786a8bdebbedae909afdf95687a254b108eb0a425e2c21814b05ee0fd53R84)

and yes (good catch:)